feat: article for 4.14 (#1209)
withastro/astro.build: src/content/blog/astro-4140.mdx

---
title: "Astro 4.14"

> description: Astro 4.14 is available now! This release includes a first experimental version of Content Layer, experimental support for intellisense inside content files, and more.

Reviewer: Yeah, I think it's a Microsft trademark!

- [**Experimental: Content Layer**](#experimental-content-layer)
- [**Experimental: Intellisense inside content**](#experimental-intellisense-inside-content)
- [**New `injectTypes` integration API**](#new-injecttypes-integration-api)
- [**Support for metastrings in the Code component**](#support-for-metastrings-in-the-code-component)

To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:

```sh
# Recommended:
npx @astrojs/upgrade

# Manual:
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## Experimental: Content Layer

In Astro 2.0 we introduced [Content Collections](https://docs.astro.build/en/guides/content-collections/), a powerful way to manage local content in Astro projects. While this has helped hundreds of thousands of developers build content-driven sites with Astro, we've heard your feedback that you want more flexibility and power in managing your content. We're excited to announce the experimental release of the next generation of content collections in Astro: the Content Layer API.

> The new Content Layer API builds upon content collections, taking them beyond local files in `src/content/` and allowing you to fetch content from anywhere, including remote APIs, or files anywhere in your project. These new collections work alongside your existing content collections, and you can migrate them to the new API at your own pace. We think you'll want to soon, because they have lots of advantages, but there's no rush – the existing collections will continue to work as they always have.

Reviewer 2: ```suggestion
The new Content Layer API builds upon content collections, taking them beyond local files in `src/content/` and allowing you to fetch content from anywhere, including remote APIs. These new collections work alongside your existing content collections, and you can migrate them to the new API at your own pace. We think you'll want to soon, because they have lots of advantages, but there's no rush – the existing collections will continue to work as they always have.
```
We made a similar edit to the documentation to tighten this up. I *could* see explicitly mentioning the file system as an alternative, as that might be considered a cool thing, too. Or, I'd let the "remote APIs" shine because people will be excited for this!

> To try out the new Content Layer API, first enable it in your Astro config:

Reviewer: Done!

```json
{
	"astro.content-intellisense": true
}
```

This feature currently only works in the editor, and will be enabled in `astro check` in a future release.

## Deprecate support for dynamic `prerender` values

In order to improve our bundling and tree-shaking capabilities, Astro 4.14 deprecates support for dynamic `prerender` values. This means that you can no longer use dynamic values for the `prerender` option in your Astro pages.

This feature can now be replaced with the new `astro:route:setup` hook in Astro integrations, which allows you to dynamically control the prerendering of specific pages.

```diff
---
- export const prerender = import.meta.env.PROD;
---
```

```js
// astro.config.mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  integrations: [setPrerender()],
});

function setPrerender() {
  return {
    name: 'set-prerender',
    hooks: {
      'astro:route:setup': ({ route }) => {
        if (route.component.endsWith('/blog/[slug].astro')) {
          route.prerender = true;
        }
      },
    },
  };
}
```

## New `injectTypes` integration API

> Type generation has become all the rage in recent years, and Astro is no exception as we use type generation pretty extensively for features like Content Collections (or, now, Content Layer), or Astro DB.

Reviewer 2: ```suggestion
Type generation has become all the rage in recent years, and Astro is no exception! We use type generation pretty extensively for features like Astro DB and Content Collections (with the new Content Layer API that you're definitely going to enable as soon as you're done reading this post).
```
Not sure we've decided to imply that "content collections are no longer a thing" but here's another way if you wanted to inject some personality into this line!

```sh
# Recommended:
npx @astrojs/upgrade

# Manual:
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## Experimental: Content Layer API

In Astro 2.0 we introduced [Content Collections](https://docs.astro.build/en/guides/content-collections/), a powerful way to manage local content in Astro projects. While this has helped hundreds of thousands of developers build content-driven sites with Astro, we've heard your feedback that you want more flexibility and power in managing your content. We're excited to announce the experimental release of the next generation of content collections in Astro: the Content Layer API.

The new Content Layer API builds upon content collections, taking them beyond local files in `src/content/` and allowing you to fetch content from anywhere, including remote APIs. These new collections work alongside your existing content collections, and you can migrate them to the new API at your own pace. We think you'll want to soon, because they have lots of advantages, but there's no rush – the existing collections will continue to work as they always have.

### Performance

> The Content Layer API is the first step in scaling Astro to support fast builds for sites with tens or hundreds of thousands of pages. The new API is designed to be more efficient and performant in the way it loads content. It caches content locally to avoid the need to keep hitting APIs, and has dramatically improved handling of local Markdown and MDX files. This benchmark shows the performance when building a site with 10,000 pages on a Macbook Air M1:

Reviewer 3: unnecessary comma
```suggestion
The Content Layer API is the first step in scaling Astro to support fast builds for sites with tens or hundreds of thousands of pages. The new API is designed to be more efficient and performant in the way it loads content. It caches content locally to avoid the need to keep hitting APIs and has dramatically improved handling of local Markdown and MDX files. This benchmark shows the performance when building a site with 10,000 pages on a Macbook Air M1:
```

  schema: z.object({
    title: z.string(),
    description: z.string(),
    publishDate: z.coerce.date(),
  })
});

export const collections = { blog };
```

You can then use the content in your site:

```astro
---
import { getEntry, render } from 'astro:content';

const post = await getEntry('blog', Astro.params.slug);

const { Content } = await render(entry);
---

<Content />
```

### Creating a loader

You're not restricted to the built-in loaders – we hope you'll try building your own. You can fetch content from anywhere and return an array of entries:

```ts
// src/content/config.ts
const countries = defineCollection({
  loader: async () => {
    const response = await fetch("https://restcountries.com/v3.1/all");
    const data = await response.json();
    // Must return an array of entries with an id property,
    // or an object with IDs as keys and entries as values
    return data.map((country) => ({
      id: country.cca3,
      ...country,
    }));
  },
  // optionally add a schema to validate the data and make it type-safe for users
  // schema: z.object...
});

export const collections = { countries };
```

> For more advanced loading logic, you can define an object loader. This allows incremental updates and conditional loading, and gives full access to the data store. It also allows a loader to define its own schema, including generating it dynamically based on the source API. See the [the Content Layer API RFC](https://github.com/withastro/roadmap/blob/content-layer/proposals/0047-content-layer.md#loaders) for more details.

Reviewer 3: I made this same suggestion in the Configuration Reference's types comments, just a small styling change to avoid multiple chained `and`s
```suggestion
For more advanced loading logic, you can define an object loader. This allows incremental updates and conditional loading while also giving full access to the data store. It also allows a loader to define its own schema, including generating it dynamically based on the source API. See the [the Content Layer API RFC](https://github.com/withastro/roadmap/blob/content-layer/proposals/0047-content-layer.md#loaders) for more details.
```

## Experimental: Intellisense inside content files

Astro 4.14 introduces experimental support for Intellisense inside the frontmatter of your content. This feature helps you write content files more efficiently by providing completions, validation, hover information, and more for frontmatter keys and values. These tools are based on your content schemas, and available directly in your editor.

This feature is available in VS Code and other editors that support the Language Server Protocol for Markdown, MDX, and Markdoc files.

<BlogContentImage src={frontmatterIntellisense} />

To enable content file intellisense, add the following to your Astro config:

```js
import { defineConfig } from 'astro';

export default defineConfig({
	experimental: {
		contentIntellisense: true
	}
})
```

and in VS Code, enable the `astro.content-intellisense` setting. For other editors, pass `contentIntellisense: true` to the initialization parameters of the Astro language server.

```json
{
	"astro.content-intellisense": true
}
```

> This feature currently only works in the editor, and will be enabled in `astro check` in a future release.

Reviewer 3: unnecessary comma
```suggestion
This feature currently only works in the editor and will be enabled in `astro check` in a future release.
```

## Deprecate support for dynamic `prerender` values

In order to improve our bundling and tree-shaking capabilities, Astro 4.14 deprecates support for using dynamic values for the `prerender` option in your Astro pages.

This feature can now be replaced with the new `astro:route:setup` hook in Astro integrations, which allows you to dynamically control the prerendering of specific pages.

> ```diff
> ---
> - export const prerender = import.meta.env.PROD;
> ---
> ```

Reviewer 3: We have Expressive Code set up here, which means it allows us to use diff syntax + syntax highlighting.
````suggestion
```diff lang="astro"
---
- export const prerender = import.meta.env.PROD;
---
```
````

      'astro:route:setup': ({ route }) => {
        if (route.component.endsWith('/blog/[slug].astro')) {
          route.prerender = true;
        }
      },
    },
  };
}
```

## New `injectTypes` integration API

Type generation has become all the rage in recent years, and Astro is no exception! We use type generation pretty extensively for features like Astro DB and Content Collections (or the new Content Layer API).

Thanks to [@florian-lefebvre](https://github.com/florian-lefebvre), Astro integrations can now join the fun with the new `injectTypes` integration API! This API allows integrations to inject types into the user's project, making it easier to provide type definitions for your integration's features.

```js
// my-integration/index.js
export default {
	name: 'my-integration',
	'astro:config:done': ({ injectTypes }) => {
		injectTypes({
			filename: "types.d.ts",
			content: "declare module 'virtual:my-integration' {}"
		})
	}
};
```

For more information on this feature, check out the [Astro integration API documentation](https://docs.astro.build/en/reference/integrations-reference/#injecttypes-options).

## Support for metastrings in the Code component

> Astro 4.14 adds support for the `meta` attribute in the `Code` component. This attribute simulate using properties on code blocks in Markdown, for example `` ```js astro=cool ``, which can then be used by [Shiki transformers](https://shiki.matsu.io/guide/transformers#meta) to apply custom transformations to the code block.

Reviewer 3: wording nit
```suggestion
Astro 4.14 adds support for the `meta` attribute in the `Code` component. This attribute simulates using properties on code blocks in Markdown, for example `` ```js astro=cool ``, which can then be used by [Shiki transformers](https://shiki.matsu.io/guide/transformers#meta) to apply custom transformations to the code block.
```
